x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -272,6 +274,59 @@ public void createUserToken(Authentication authentication, Authentication origin
         }
     }
 
+    /**
+     * Asynchronously creates a new API key based off of the request and authentication
+     * @param authentication the authentication that this api key should be based off of
+     * @param request the request to create the api key included any permission restrictions
+     * @param listener the listener that will be used to notify of completion
+     */
+    public void createApiKey(Authentication authentication, CreateApiKeyRequest request, ActionListener<CreateApiKeyResponse> listener) {
```

Comment:
The TokenService is already quite large and hard to reason about. I would love to see it broken up at some point during the development of this feature.